src/custom-element.js

```diff
@@ -1619,137 +1588,47 @@ function createBaseCustomElementClass(win) {
-        const doc = this.ownerDocument;
-        devAssert(doc);
-
-        const container = htmlFor(/** @type {!Document} */ (doc))`
-            <div class="i-amphtml-loading-container i-amphtml-fill-content
-              amp-hidden"></div>`;
-        const loadingElement = createLoaderElement(
-          this.getAmpDoc(),
-          this,
-          this.layoutWidth_,
-          this.layoutHeight_,
-          startTime
-        );
-
-        container.appendChild(loadingElement);
-
-        this.appendChild(container);
-        this.loadingContainer_ = container;
-        this.loadingElement_ = loadingElement;
-      }
-    }
-
     /**
      * Turns the loading indicator on or off.
      * @param {boolean} state
-     * @param {{cleanup:(boolean|undefined), startTime:(number|undefined)}=} opt_options
+     * @param {boolean=} ignoreLoadingState
      * @public @final
      */
-    toggleLoading(state, opt_options) {
-      const cleanup = opt_options && opt_options.cleanup;
-      const startTime = opt_options && opt_options.startTime;
-      assertNotTemplate(this);
-
-      if (state === this.loadingState_ && !opt_options) {
-        // Loading state is the same.
-        return;
-      }
-      this.loadingState_ = state;
-      if (!state && !this.loadingContainer_) {
-        return;
-      }
-
-      // Check if loading should be shown.
-      if (state && !this.isLoadingEnabled_()) {
-        this.loadingState_ = false;
+    toggleLoading(state, ignoreLoadingState = false) {
+      // TODO(dvoytenko, #9177): cleanup `this.ownerDocument.defaultView`
+      // once investigation is complete. It appears that we get a lot of
+      // errors here once the iframe is destroyed due to timer.
+      if (!this.ownerDocument || !this.ownerDocument.defaultView) {
         return;
       }
 
-      this.mutateOrInvoke_(
-        () => {
-          let state = this.loadingState_;
-          // Repeat "loading enabled" check because it could have changed while
-          // waiting for vsync.
-          if (state && !this.isLoadingEnabled_()) {
-            state = false;
-          }
-          if (state) {
-            this.prepareLoading_(startTime);
-          }
-          if (!this.loadingContainer_) {
-            return;
-          }
-
-          this.loadingContainer_.classList.toggle('amp-hidden', !state);
-          this.loadingElement_.classList.toggle('amp-active', state);
-
-          if (!state && cleanup && !this.implementation_.isLoadingReused()) {
-            const loadingContainer = this.loadingContainer_;
-            this.loadingContainer_ = null;
-            this.loadingElement_ = null;
-            this.mutateOrInvoke_(
-              () => {
-                dom.removeElement(loadingContainer);
-              },
-              undefined,
-              true
-            );
-          }
-        },
-        undefined,
-        /* skipRemeasure */ true
+      const loadingIndicator = Services.loadingIndicatorOrNull(
```

Comment:
nit: don't we usually allow the passing of an `Element` here? E.g. `this.element_`

Reply:
Either or is ok. The `ampdoc` is faster, b/c when an element is provided, it's first resolved to the `ampdoc` and then the service is looked up. So we are reusing the `ampdoc` lookup here done before.